step.yml

```diff
@@ -1,12 +1,33 @@
 title: "Git Clone Repository"
-summary: Clone a repository to the specified path on the VM
+summary: Clones your source code from a remote repository into a clean build machine on Bitrise.
 description: |
-  Clones a repository based on the provided inputs.
+   It runs `git clone` command to pull source code from a remote repository into the build machine on Bitrise. When triggering a build or starting a build manually, a webhook notifies Bitrise which branch and commit should be cloned for the build and passes it onto the **Git Clone Repository** Step. Once the build has run, the build machine gets rolled back so no source code is stored on Bitrise.
```

Comment:
It runs `git clone` command to pull source code from a remote repository into the build machine on Bitrise. When triggering a build or starting a build manually, a webhook notifies Bitrise which branch and commit should be cloned for the build and passes it onto the **Git Clone Repository** Step. Once the build has run, the build machine gets rolled back so no source code is stored on Bitrise.

->

The step checks out the defined repository state, optionally updates the repository submodules and exports the achieved git repository state properties.

_If the `clone_into_dir` is an empty directory the step initializes a git repository and adds the defined remote `repository_url` before checkout._

## Checkout

The checkout process depends on the checkout properties: the step either checks out a repository state defined by a git commit or git tag or achieves a merged state of a Pull / Merge request.

The step uses two solutions to achieve the merged state of the Pull / Merge request: auto merge in case merge branch or diff file provided by the Git service and manual merge otherwise.

## Auto merge

The auto merge process: fetches the Pull / Merge request destination branch

in case of merge branch: fetches the merge branch, pulls the destination branch and merges the merge branch into the destination branch.

in case of diff file: downloads the diff file and applies it as a patch.

## Manual merge

The manual merge process: fetches and pulls the Pull / Merge request destination branch

in case of the Pull / Merge request source is a forked repository: adds a forked repo as another remote, fetches this remote too and merges the fork into the destination branch

otherwise fetches the Pull / Merge request source and merges the source branch into the destination branch.

## After checkout

After the desired state is checked out, the step optionally updates the submodule, in case of Pull / Merge request, checks out a detach head and exports the achieved git state properties.